SI-7247, deprecate NotNull.

paulp opened the pull request:
Brought the steamroller to bear on NotNull; fired up the Zamboni afterward to smooth things over.

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -462,12 +461,9 @@ trait Types
-    def notNull: Type =
-      if (!settings.Ynotnull.value || isNotNull || phase.erasedTypes) this
-      else NotNullType(this)
+    @deprecated("This method will be removed", "2.11.0") def notNull: Type = this
```
we're in reflect.internal -- we can probably get rid of this as well

paulp replied:
Would be glad to dump it, but I'm always breaking the IDE and I figured a one line identity function was cheap insurance.

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2347,8 +2333,7 @@ trait Types
 
-    override def isNotNull =
-      sym.isModuleClass || sym == NothingClass || (sym isNonBottomSubClass NotNullClass) || super.isNotNull
+    override def isNotNull = sym == NothingClass || super.isNotNull
```
is this still referenced?

paulp replied:
isNotNull is not used much if at all, but if you read my commit messages (github seems to discourage this, lame) you will see it was intentional.

paulp replied:
You know what: IAGNI. You are right, enjoy this moment. I'll remove it.

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4080,8 +4065,7 @@ trait Types
 
   protected[internal] def containsNull(sym: Symbol): Boolean =
```
also unreferenced, I think

paulp replied:
containsNull is used in subtyping - Null <:< X if containsNull(X).

adriaanm commented:
Great. I almost finished cobbling together my own PR when I checked email and saw this. This LBTM (think "better", not "lobotomy") once we prune those last unused things, unless they are used?

paulp commented:
Will rebase/remove/reopen.

paulp committed:
Removed -Ynotnull setting.

It never worked and it would periodically jump out and bite
someone. Slash and burn so new plants can take root. Eliminated
NotNullType and NotNullClass, internal elements. Removed notNull
method on Type, which no longer has a purpose. Retained 'def
isNotNull' because there are types which are known by construction
not to be null (ThisType, ConstantType as long as the constant
isn't null, etc.) and that's still useful to know.

paulp committed:
SI-7247, deprecated NotNull.

Removed NotNull from tests and the parentage of AnyVal.

Removed the tests which were actually testing anything to
do with NotNull; massaged the others to forget NotNull and/or
not to name local things NotNull.

paulp committed:
Completely remove isNotNull/notNull.

Restarted the Zamboni and collected these as well.
Author was convinced by reviewer that clinging to
isNotNull like Linus's security blanket will not
help us with landing a picture of the Great Pumpkin.
YAGNI, Charlie Brown.

paulp committed:
Remove -Xcheck-null setting.

Maybe this was useful in some way; but no way I ever saw.
I have comments which tell me this is exposed in the IDE so
I left a stub.

I also removed mkCheckInit. That probably sounds like it's related
to -Xcheckinit. Guess again, guy-who-thinks-names-mean-things. It
was only used by -Xcheck-null.

paulp committed:
Eliminated containsNull.

This was a little trickier than the previous. I introduced
a new method 'isBottomSubClass' which is the obvious complement
to the beloved 'isNonBottomSubClass'. In eliminating the two
call sites of containsNull I might have overshot the mark a
bit when I rewrote fourthTry and thirdTryRef, but who is going
to argue with such beauty as this.

paulp commented:
Okay @adriaanm this should now be worthy of your attention.

adriaanm commented:
I hope so.

paulp commented:
@adriaanm, perhaps it's time for YOU to try to be worthy of IT. Too busy making smoothies, I bet.

adriaanm commented:
LGTM

adriaanm commented:
Great JIRA is "temporarily unavailable" (I wish it was permanent). Ticket still needs to be closed.

lrytz commented:
i started it again

cchantep commented:
If I properly get it, if a value whose type is currently marked with this trait is given a `null`, a compilation error is raised (scala version = 2.10.4-RC2, using SBT 0.13.1):

```
type mismatch;
[error]  found   : Null(null)
[error]  required: ANotNullType
```

What will replacement for such case, either for scala-lib type (e.g. `Int`, `Float`, ...) or those outside?

adriaanm commented:
NotNull only worked in trivial cases and it proved hard to extend it beyond
them in a useful way, so we'd rather not have it at all.

On Saturday, March 15, 2014, cchantep notifications@github.com wrote:

> If I properly get it, if a value whose type is currently marked with this
> trait is given a null, a compilation error is raised (scala version =
> 2.10.4-RC2, using SBT 0.13.1):
> 
> type mismatch;
> [error]  found   : Null(null)
> [error]  required: ANotNullType
> 
> What will replacement for such case, either for scala-lib type (e.g. Int,
> Float, ...) or those outside?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2244#issuecomment-37725193
> .

cchantep commented:
I can understand, but that's a quite radical pov.
